include/unifex/detach_on_cancel.hpp

```diff
@@ -0,0 +1,172 @@
+      unifex::start(childOp);
+    }
+
+    DownstreamReceiver receiver_;
+    manual_lifetime<typename stop_token_type_t<
+        DownstreamReceiver>::template callback_type<cancel_callback>>
+        callback_;
+    std::unique_ptr<detached_state> state_;
+  };
+};
+
+template <typename UpstreamSender, typename DownstreamReceiver>
+struct operation_state<UpstreamSender, DownstreamReceiver>::_receiver final {
+  auto tryGetOp() noexcept {
+    // if cancellation won, parentOp_ would be null at this point
+    // see line24
+    auto* op = state_->parentOp_.exchange(nullptr);
+    if (op != nullptr) {
+      op->callback_.destruct();
+    } else {
+      delete state_;
+    }
+    return op;
+  }
+
+  template <typename... Values>
+  void set_value(Values&&... values) noexcept {
+    if (auto op = tryGetOp()) {
+      UNIFEX_TRY { unifex::set_value(std::move(op->receiver_), (Values &&) values...); }
+      UNIFEX_CATCH(...) {
+        unifex::set_error(std::move(op->receiver_), std::current_exception());
+      }
+    }
+  }
+
+  template <typename Error>
+  void set_error(Error&& error) noexcept {
+    if (auto op = tryGetOp()) {
+      unifex::set_error(std::move(op->receiver_), (Error &&) error);
+    }
+  }
+
+  void set_done() noexcept {
+    if (auto op = tryGetOp()) {
+      unifex::set_done(std::move(op->receiver_));
+    }
+  }
+
+  friend inplace_stop_token
+  tag_invoke(tag_t<get_stop_token>, const _receiver& r) noexcept {
+    return r.state_->stopSource_.get_token();
+  }
+  detached_state* state_;
+};
+
+template <typename UpstreamSender, typename DownstreamReceiver>
+struct operation_state<UpstreamSender, DownstreamReceiver>::detached_state final {
+    detached_state(typename operation_state<UpstreamSender, DownstreamReceiver>::type* op, UpstreamSender&& s)
+      noexcept(is_nothrow_connectable_v<UpstreamSender, _receiver>)
+  : parentOp_(op)
+  , childOp_(connect(std::move(s), _receiver{this})){};
+
+
+  std::atomic<typename operation_state<UpstreamSender, DownstreamReceiver>::type*> parentOp_;
+  inplace_stop_source stopSource_;
+  connect_result_t<UpstreamSender, _receiver> childOp_;
+};
+
+template <typename Sender>
+struct _sender {
+  struct type;
+};
+
+template <typename Sender>
+using sender = typename _sender<Sender>::type;
+
+template <typename Sender>
+struct _sender<Sender>::type {
+  UNIFEX_NO_UNIQUE_ADDRESS Sender upstreamSender_;
+
+  template <
+      template <typename...>
+      class Variant,
+      template <typename...>
+      class Tuple>
+  using value_types =
+      typename sender_traits<Sender>::template value_types<Variant, Tuple>;
+
+  template <template <typename...> class Variant>
+  using error_types =
+      typename sender_traits<Sender>::template error_types<Variant>;
+
+  static constexpr bool sends_done = true;
+
+  friend constexpr auto
+  tag_invoke(tag_t<blocking>, const type& sender) noexcept {
+    return blocking(sender.upstreamSender_);
+  }
```

Comment:
I think this can potentially complete inline in the case that stop has been requested.
So if `blocking(upstreamSender)` returns `never` then you should return `maybe`, otherwise you can just return `blocking(upstreamSender)`.